add label generation api

jake-walker opened the pull request:
<!--
  This template provides some ideas of things to include in your PR description.
  To start, try providing a short summary of your changes in the Title above.
  If a section of the PR template does not apply to this PR, then delete that section.
 -->

This PR adds a label generator which can generate labels for items (either by ID or asset ID) and locations. In the config, a command can be set for printing labels out, the idea is that this would be you can pass the generated filename to a command that will print out the label. Alternatively, users can simply download the label through the UI to print another way.

_This is just a draft for now as I'd like some feedback and still need to make this work nicely with the existing label generation "report" page._

![Screenshot of new label dropdown on an item](https://github.com/user-attachments/assets/aafd2979-7a75-4a5b-927a-0228d376d6fb)

**Print confirmation modal**
![](https://github.com/user-attachments/assets/f8e8dcfa-cd76-46bd-ba9c-25fa69fc513d)

**Location label**
![Location label example](https://github.com/user-attachments/assets/d5eff936-7ccf-429c-819f-3e558802e2da)

**Item label**
![Item label example](https://github.com/user-attachments/assets/54d7be74-4a9f-4ab4-9382-50d4b8e7a116)

**Asset label**
![Asset label example](https://github.com/user-attachments/assets/561d71f1-b3c8-4ad6-8d5b-69b08cbc847b)

## What type of PR is this?

<!--
  Delete any of the following that do not apply:
 -->

- feature

## What this PR does / why we need it:

<!--
  What goal is this change working towards?
  Provide a bullet pointed summary of how each file was changed.
  Briefly explain any decisions you made with respect to the changes.
  Include anything here that you didn't include in *Release Notes*
  above, such as changes to CI or changes to internal methods.
-->

There are a bunch of people who want to be able to add labels to their physical Homebox locations, myself included, so this adds that functionality.

- A label button has been added to item & location pages
- Generated label width, height, font size and padding can be configured in the application config
- A command can be set in the config to print labels remotely from the UI

## Which issue(s) this PR fixes:

<!--
If this PR fixes one of more issues, list them here.
One line each, like so:
Fixes #123
Fixes #39
-->

Resolves #9 (now part of #436)

## Special notes for your reviewer:

This is just a draft for now, I'm planning to give this a tidy up, and make sure it integrates nicely with the existing label generation page. In addition, I'd like to know any other issues I could potentially include (e.g. #495) or any feedback at all on this feature.

<!--
   Is there any particular feedback you would / wouldn't like?
   Which parts of the code should reviewers focus on?
-->

## Testing

This has just been manually tested, but I can add tests if needed.

<!--
  Describe how you tested this change.
-->

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

Based on the comprehensive pull request summary, here are the updated release notes:

- **New Features**
  - Added label generation and printing functionality for items, locations, and assets.
  - Introduced new API endpoints for retrieving labels with optional print functionality.
  - Implemented a new Vue component for label management with print and download options.
  - Added a QR code scanning feature in the navigation.

- **Configuration**
  - Added new configuration options for label dimensions, padding, font size, and print commands.

- **User Interface**
  - Integrated label maker component into item and location detail pages.
  - Enhanced navigation with a new scanner route.

- **API Enhancements**
  - Created new `/labelmaker` endpoints for generating labels.
  - Updated API summary to include label printing status.

These changes expand Homebox's labeling capabilities, providing users with more flexible options for generating and managing item, location, and asset labels.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

jake-walker commented:
Yeah I like that idea of just referencing an ID or something, it makes the QR codes smaller too, but maybe I could have a look at adding an option for it so it can be turned on or off.

fidoriel commented:
Then scanning in the ui is required to support it. What about the location? I would like to be able to have a sticker on the tools where it says this belongs into this box. Example: Item: Wera Screwdriver PZ 2 only asset from its kind item. Located in toolbox 42. I would like to have the option to print QR Code, Name/ID and `location/item/asset` on the label to know where to put it back, or for other people to know.

jake-walker commented:
That'd be great, I've not started using HomeBox yet so I wouldn't have thought of that but I'll add it in 🙂

fidoriel commented:
> That'd be great, I've not started using HomeBox yet so I wouldn't have thought of that but I'll add it in 🙂

Printing is the single killer feature for me, since I have got a printer the effort seems reasonable to me. Without a printer and Homebox not supporting it, it seemed pointless/too much effort to me.
I am also willing to contribute with my rusted go experience. Or if I can test with my Brother QL 800 within different OS I will happily do it

jake-walker commented:
I've updated the label generation code to add the location name in the description:

![Example Label](https://github.com/user-attachments/assets/139ec90a-42e0-4fc9-9970-3180c53cd82e)

And added a new page where you can scan labels. I've not adjusted the labels to _only_ be IDs, because I think it's probably quite useful to be able to scan them with other things, for example, your phone. But this scan page will ignore the start of the URL so if you were to change the server URL the barcodes would still scan and navigate to the right place from this page.

![Scan Page](https://github.com/user-attachments/assets/dad3cdea-769e-48c9-8e7e-b0f2b6b6faaf)

jake-walker commented:
> I am also willing to contribute with my rusted go experience. Or if I can test with my Brother QL 800 within different OS I will happily do it

That would be great. I've not got a label printer I can automate very easily, so it would be super useful to have someone test it out. I assume it's easy to link up with the command interface I added, but it seems potentially a bit flawed. And I'm certainly open to help on this :)

fidoriel commented:
> > I am also willing to contribute with my rusted go experience. Or if I can test with my Brother QL 800 within different OS I will happily do it
> 
> That would be great. I've not got a label printer I can automate very easily, so it would be super useful to have someone test it out. I assume it's easy to link up with the command interface I added, but it seems potentially a bit flawed. And I'm certainly open to help on this :)

Tell me what to do.

fidoriel commented:
Today I took the time to try it out.

There was a misconception on my side: My Brother QL 800 is connected to my Windows Workshop PC and not my homeserver.
Label downloading works fine.
I tried printing it with the browser print dialog. After changing print settings, Disable `Header and Footer`, enable `scale to fit` choosing size `62x29mm` and `Portrait` as well as `Margins None` I was able to print the label and the printer did cut.

I would like to see no margins in the image, so I can set them on the print dialog and a bit thicker font. Also, I was wondering about why jpeg is used and not png, but the jpeg artifacts seem not to matter in this context.

I added this to print from dialog.
```vue
function openBrowserPrintDialog() {
  const printWindow = window.open(getLabelUrl(false));
  printWindow.onload = function() {
    printWindow.print();
  };
}


<li>
  <button @click="openBrowserPrintDialog">
    <MdiPrinterPos name="mdi-printer-pos" class="mr-2" />
    {{ $t("components.global.label_maker.print_browser") }}
  </button>
</li>
```


![photo_2025-02-08_16-08-03](https://github.com/user-attachments/assets/71e8b92a-0da2-457d-9cb9-68071bd3f99b)

jake-walker commented:
> Today I took the time to try it out.
> 
> There was a misconception on my side: My Brother QL 800 is connected to my Windows Workshop PC and not my homeserver. Label downloading works fine. I tried printing it with the browser print dialog. After changing print settings, Disable `Header and Footer`, enable `scale to fit` choosing size `62x29mm` and `Portrait` as well as `Margins None` I was able to print the label and the printer did cut. The printers are some kind of physical limited to margins at top and bottom, so it does not make sense to add them in the image.
> 
> I would like to see no margins in the image, so I can set them on the print dialog and a bit thicker font, to max out the printers ability while wasting as little tape as possible. Also, I was wondering about why jpeg is used and not png, but the jpeg artifacts seem not to matter in this context.
> 
> I added this to print from dialog.
> 
> ```vue
> function openBrowserPrintDialog() {
>   const printWindow = window.open(getLabelUrl(false));
>   printWindow.onload = function() {
>     printWindow.print();
>   };
> }
> 
> 
> <li>
>   <button @click="openBrowserPrintDialog">
>     <MdiPrinterPos name="mdi-printer-pos" class="mr-2" />
>     {{ $t("components.global.label_maker.print_browser") }}
>   </button>
> </li>
> ```
> 
> It would be also nice if there could be a free text field on the labels so that I can enter my surname and contact like a company ;) But I can create prs on my own to build this...

Awesome, glad it worked for you. I've added in the ability to have 0 margins around the edge of the label (by setting argument `--label-maker-margin=0` or environment variable `HBOX_LABEL_MAKER_MARGIN=0`) and thickened up the font weight for the description - although the font that's being used can't go any thicker than it already is.

I think saving to a jpeg was part of the example code I was using, but I've switched it over to png, which makes more sense as you say

fidoriel commented on docs/en/configure-homebox.md:
```diff
@@ -26,6 +26,12 @@
+| HBOX_LABEL_MAKER_PADDING             | 8                                          | space between elements on label                                                        |
+| HBOX_LABEL_MAKER_MARGIN              | 8                                          | space between the label content and edges of the label                                 |
+| HBOX_LABEL_MAKER_FONT_SIZE           | 32.0                                       | the size of the labels font                                                            |
+| HBOX_LABEL_MAKER_PRINT_COMMAND       |                                            | the command to use for printing labels. if empty, label printing is disabled. `{{.FileName}}` in the command will be replaced with the jpg filename of the label |
```
@jake-walker here is still jpg in ^^

jake-walker replied:
Good spot!

fidoriel commented:
Thank you so much. But for some reason in the `scanner.vue` file, I cannot build the Docker Image.

```
8.096  ERROR  x Build failed in 3.79s
8.096 
8.097 
8.097  ERROR  Nuxt Build Error: [vite]: Rollup failed to resolve import "#components" from "/app/pages/scanner.vue?vue&type=script&setup=true&lang.ts".
8.097 This is most likely unintended because it can break your application at runtime.
8.097 If you do want to externalize this module explicitly add it to
8.097 build.rollupOptions.external
8.097 
8.097     This is most likely unintended because it can break your application at runtime.
8.097     If you do want to externalize this module explicitly add it to
8.097     build.rollupOptions.external
8.097     at viteWarn (node_modules/.pnpm/vite@5.4.11_@types+node@22.10.5_terser@5.37.0/node_modules/vite/dist/node/chunks/dep-CB_7IfJ-.js:65589:17)
8.097     at onRollupWarning (node_modules/.pnpm/vite@5.4.11_@types+node@22.10.5_terser@5.37.0/node_modules/vite/dist/node/chunks/dep-CB_7IfJ-.js:65621:5)
8.097     at onwarn (node_modules/.pnpm/vite@5.4.11_@types+node@22.10.5_terser@5.37.0/node_modules/vite/dist/node/chunks/dep-CB_7IfJ-.js:65284:7)
8.097     at node_modules/.pnpm/rollup@4.29.1/node_modules/rollup/dist/es/shared/node-entry.js:19537:13
8.097     at Object.logger [as onLog] (node_modules/.pnpm/rollup@4.29.1/node_modules/rollup/dist/es/shared/node-entry.js:21263:9)
8.097     at ModuleLoader.handleInvalidResolvedId (node_modules/.pnpm/rollup@4.29.1/node_modules/rollup/dist/es/shared/node-entry.js:20152:26)
8.097     at node_modules/.pnpm/rollup@4.29.1/node_modules/rollup/dist/es/shared/node-entry.js:20110:26
8.097 
8.142  ELIFECYCLE  Command failed with exit code 1.
```

jake-walker commented:
Sorted, looks like I just missed that when rebasing onto the main branch earlier

fidoriel commented:
> Sorted, looks like I just missed that when rebasing onto the main branch earlier

Thank you. The settings now allow configuring everything printer location, label settings and so on.

What about the optinal ability to set a branding string on every label. Something I can freely add once and then printed on each label. For example, Name and Phone?

agryson commented:
Hello @jake-walker 👋
Sorry to dig up this PR, but I was wondering if you could give an example of how the print command works?
I have set the variable and the same command works from the host terminal, but[ I am getting errors in the logs](https://github.com/sysadminsmedia/homebox/discussions/703#discussioncomment-13196470) when trying to print from server?

agryson commented:
> Hello @jake-walker 👋 Sorry to dig up this PR, but I was wondering if you could give an example of how the print command works? I have set the variable and the same command works from the host terminal, but[ I am getting errors in the logs](https://github.com/sysadminsmedia/homebox/discussions/703#discussioncomment-13196470) when trying to print from server?

Worked around this using cp, a mounted folder and a service to push changed files to the printer

jake-walker commented:
@agryson sorry for the delay. Based on what I've read on the other issue too, you'd probably need to create a custom image based on homebox with your printer CLI installed, then you can mount your printer into the container. You should also be able to print normally through the web interface without any configuration too if you've got CUPS setup

agryson commented:
Oh, print from browser always worked great, but from mobile, the print from server is ideal.
I managed to get it working with a workaround by `cp`ing to a mounted directory and a host service that watches for changes and prints. 
Would be smoother if there were an option to add `lp` or something to the container's path but not sure how easy/secure that is.
But point is, while I had to work around a few things, it is now working as I'd hoped 🎉
